File: source/nxrender/nxrender-widgets/src/controls/textfield.rs
//! TextField Widget
//!
//! A text input widget with editing support.

use crate::base::{Widget, WidgetId, EventResult, Constraints, WidgetState, Event, Key, MouseButton};
use nxgfx::{Rect, Size, Point, Color};
use nxrender_core::Painter;

/// TextField style configuration
#[derive(Debug, Clone)]
pub struct TextFieldStyle {
    /// Background color
    pub background: Color,
    /// Border color when normal
    pub border_color: Color,
    /// Border color when focused
    pub border_color_focused: Color,
    /// Border width
    pub border_width: f32,
    /// Border radius
    pub border_radius: f32,
    /// Text color
    pub text_color: Color,
    /// Placeholder color
    pub placeholder_color: Color,
    /// Cursor color
    pub cursor_color: Color,
    /// Selection background
    pub selection_color: Color,
    /// Padding
    pub padding: f32,
    /// Font size
    pub font_size: f32,
}

impl Default for TextFieldStyle {
    fn default() -> Self {
        Self {
            background: Color::WHITE,
            border_color: Color::rgb(209, 213, 219),
            border_color_focused: Color::rgb(59, 130, 246),
            border_width: 1.0,
            border_radius: 6.0,
            text_color: Color::rgb(17, 24, 39),
            placeholder_color: Color::rgb(156, 163, 175),
            cursor_color: Color::rgb(59, 130, 246),
            selection_color: Color::rgba(59, 130, 246, 40),
            padding: 10.0,
            font_size: 14.0,
        }
    }
}

/// TextField widget
pub struct TextField {
    id: WidgetId,
    text: String,
    placeholder: String,
    bounds: Rect,
    state: WidgetState,
    style: TextFieldStyle,
    cursor_pos: usize,
    selection_start: Option<usize>,
    on_change: Option<Box<dyn FnMut(&str) + Send + Sync>>,
    on_submit: Option<Box<dyn FnMut(&str) + Send + Sync>>,
    /// Maximum character length (0 = unlimited)
    max_length: usize,
    /// Is password field
    is_password: bool,
}

impl TextField {
    /// Create a new text field
    pub fn new() -> Self {
        Self {
            id: WidgetId::new(),
            text: String::new(),
            placeholder: String::new(),
            bounds: Rect::ZERO,
            state: WidgetState::new(),
            style: TextFieldStyle::default(),
            cursor_pos: 0,
            selection_start: None,
            on_change: None,
            on_submit: None,
            max_length: 0,
            is_password: false,
        }
    }
    
    /// Set placeholder text
    pub fn placeholder(mut self, placeholder: impl Into<String>) -> Self {
        self.placeholder = placeholder.into();
        self
    }
    
    /// Set initial text
    pub fn text(mut self, text: impl Into<String>) -> Self {
        self.text = text.into();
        self.cursor_pos = self.text.len();
        self
    }
    
    /// Set the style
    pub fn style(mut self, style: TextFieldStyle) -> Self {
        self.style = style;
        self
    }
    
    /// Set change handler
    pub fn on_change<F: FnMut(&str) + Send + Sync + 'static>(mut self, handler: F) -> Self {
        self.on_change = Some(Box::new(handler));
        self
    }
    
    /// Set submit handler (Enter key)
    pub fn on_submit<F: FnMut(&str) + Send + Sync + 'static>(mut self, handler: F) -> Self {
        self.on_submit = Some(Box::new(handler));
        self
    }
    
    /// Set max length
    pub fn max_length(mut self, len: usize) -> Self {
        self.max_length = len;
        self
    }
    
    /// Set as password field
    pub fn password(mut self) -> Self {
        self.is_password = true;
        self
    }
    
    /// Get the current text
    pub fn get_text(&self) -> &str {
        &self.text
    }
    
    /// Set the text
    pub fn set_text(&mut self, text: impl Into<String>) {
        self.text = text.into();
        self.cursor_pos = self.text.len().min(self.cursor_pos);
    }
    
    /// Get display text (masked for password)
    fn display_text(&self) -> String {
        if self.is_password {
            "•".repeat(self.text.len())
        } else {
            self.text.clone()
        }
    }
    
    /// Insert character at cursor
    fn insert_char(&mut self, c: char) {
        if self.max_length > 0 && self.text.len() >= self.max_length {
            return;
        }
        self.text.insert(self.cursor_pos, c);
        self.cursor_pos += 1;
    }
    
    /// Delete character before cursor
    fn backspace(&mut self) {
        if self.cursor_pos > 0 {
            self.cursor_pos -= 1;
            self.text.remove(self.cursor_pos);
        }
    }
    
    /// Delete character at cursor
    fn delete(&mut self) {
        if self.cursor_pos < self.text.len() {
            self.text.remove(self.cursor_pos);
        }
    }
    
    /// Get current border color
    fn current_border_color(&self) -> Color {
        if self.state.focused {
            self.style.border_color_focused
        } else {
            self.style.border_color
        }
    }
}

impl Default for TextField {
    fn default() -> Self {
        Self::new()
    }
}

impl Widget for TextField {
    fn id(&self) -> WidgetId {
        self.id
    }
    
    fn type_name(&self) -> &'static str {
        "TextField"
    }
    
    fn bounds(&self) -> Rect {
        self.bounds
    }
    
    fn set_bounds(&mut self, bounds: Rect) {
        self.bounds = bounds;
    }
    
    fn measure(&self, constraints: Constraints) -> Size {
        let min_width = 100.0;
        let height = self.style.font_size + self.style.padding * 2.0;
        
        constraints.constrain(Size::new(min_width, height))
    }
    
    fn render(&self, painter: &mut Painter) {
        if !self.state.visible {
            return;
        }
        
        // Background
        painter.fill_rounded_rect(
            self.bounds,
            self.style.background,
            self.style.border_radius,
        );
        
        // Border
        painter.stroke_rounded_rect(
            self.bounds,
            self.current_border_color(),
            self.style.border_width,
            self.style.border_radius,
        );
        
        // Text or placeholder
        let text_x = self.bounds.x + self.style.padding;
        let text_y = self.bounds.y + (self.bounds.height - self.style.font_size) / 2.0;
        
        if self.text.is_empty() && !self.placeholder.is_empty() {
            painter.draw_text(
                &self.placeholder,
                Point::new(text_x, text_y),
                self.style.placeholder_color,
            );
        } else {
            painter.draw_text(
                &self.display_text(),
                Point::new(text_x, text_y),
                self.style.text_color,
            );
        }
        
        // Cursor (if focused)
        if self.state.focused {
            let cursor_x = text_x + (self.cursor_pos as f32 * self.style.font_size * 0.6);
            let cursor_y = self.bounds.y + self.style.padding / 2.0;
            let cursor_height = self.bounds.height - self.style.padding;
            
            painter.fill_rect(
                Rect::new(cursor_x, cursor_y, 2.0, cursor_height),
                self.style.cursor_color,
            );
        }
    }
    
    fn handle_event(&mut self, event: &Event) -> EventResult {
        if !self.state.enabled {
            return EventResult::Ignored;
        }
        
        match event {
            Event::MouseDown { button: MouseButton::Left, pos } => {
                if self.bounds.contains(*pos) {
                    // Calculate cursor position from click
                    let text_x = self.bounds.x + self.style.padding;
                    let click_offset = pos.x - text_x;
                    let char_width = self.style.font_size * 0.6;
                    let new_pos = (click_offset / char_width).round() as usize;
                    self.cursor_pos = new_pos.min(self.text.len());
                    
                    EventResult::NeedsRedraw
                } else {
                    EventResult::Ignored
                }
            }
            Event::FocusGained => {
                self.state.focused = true;
                EventResult::NeedsRedraw
            }
            Event::FocusLost => {
                self.state.focused = false;
                self.selection_start = None;
                EventResult::NeedsRedraw
            }
            Event::TextInput { text } if self.state.focused => {
                for c in text.chars() {
                    self.insert_char(c);
                }
                if let Some(handler) = &mut self.on_change {
                    handler(&self.text);
                }
                EventResult::NeedsRedraw
            }
            Event::KeyDown { key, .. } if self.state.focused => {
                match key {
                    Key::Backspace => {
                        self.backspace();
                        if let Some(handler) = &mut self.on_change {
                            handler(&self.text);
                        }
                        EventResult::NeedsRedraw
                    }
                    Key::Delete => {
                        self.delete();
                        if let Some(handler) = &mut self.on_change {
                            handler(&self.text);
                        }
                        EventResult::NeedsRedraw
                    }
                    Key::Left => {
                        if self.cursor_pos > 0 {
                            self.cursor_pos -= 1;
                        }
                        EventResult::NeedsRedraw
                    }
                    Key::Right => {
                        if self.cursor_pos < self.text.len() {
                            self.cursor_pos += 1;
                        }
                        EventResult::NeedsRedraw
                    }
                    Key::Home => {
                        self.cursor_pos = 0;
                        EventResult::NeedsRedraw
                    }
                    Key::End => {
                        self.cursor_pos = self.text.len();
                        EventResult::NeedsRedraw
                    }
                    Key::Enter => {
                        if let Some(handler) = &mut self.on_submit {
                            handler(&self.text);
                        }
                        EventResult::Handled
                    }
                    _ => EventResult::Ignored,
                }
            }
            _ => EventResult::Ignored,
        }
    }
    
    fn state(&self) -> WidgetState {
        self.state
    }
    
    fn can_focus(&self) -> bool {
        self.state.enabled
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    
    #[test]
    fn test_textfield_creation() {
        let tf = TextField::new().placeholder("Enter text...");
        assert!(tf.get_text().is_empty());
    }
    
    #[test]
    fn test_textfield_insert() {
        let mut tf = TextField::new();
        tf.insert_char('H');
        tf.insert_char('i');
        assert_eq!(tf.get_text(), "Hi");
    }
    
    #[test]
    fn test_textfield_backspace() {
        let mut tf = TextField::new().text("Hello");
        tf.backspace();
        assert_eq!(tf.get_text(), "Hell");
    }
}

File: source/nxrender-cpp/widgets/textfield.cpp
// Copyright (c) 2025 KetiveeAI. All rights reserved.
// Licensed under KPL-2.0. See LICENSE file for details.
/**
 * @file textfield.cpp
 * @brief Text input field implementation
 */

#include "widgets/textfield.h"
#include "nxgfx/context.h"
#include "theme/theme.h"
#include <algorithm>

namespace NXRender {

TextField::TextField() : placeholder_("") {
    setPadding(EdgeInsets(8, 12));
    setBackgroundColor(Color::white());
}

TextField::TextField(const std::string& placeholder) 
    : placeholder_(placeholder) {
    setPadding(EdgeInsets(8, 12));
    setBackgroundColor(Color::white());
}

TextField::~TextField() = default;

void TextField::setValue(const std::string& value) {
    value_ = value;
    if (maxLength_ > 0 && static_cast<int>(value_.length()) > maxLength_) {
        value_ = value_.substr(0, maxLength_);
    }
    cursorPos_ = std::min(cursorPos_, static_cast<int>(value_.length()));
    if (changeHandler_) {
        changeHandler_(value_);
    }
}

void TextField::setSelection(int start, int end) {
    selectionStart_ = std::clamp(start, 0, static_cast<int>(value_.length()));
    selectionEnd_ = std::clamp(end, 0, static_cast<int>(value_.length()));
}

void TextField::selectAll() {
    selectionStart_ = 0;
    selectionEnd_ = static_cast<int>(value_.length());
}

std::string TextField::selectedText() const {
    if (selectionStart_ == selectionEnd_) return "";
    int start = std::min(selectionStart_, selectionEnd_);
    int end = std::max(selectionStart_, selectionEnd_);
    return value_.substr(start, end - start);
}

void TextField::setCursorPosition(int pos) {
    cursorPos_ = std::clamp(pos, 0, static_cast<int>(value_.length()));
}

void TextField::render(GpuContext* ctx) {
    if (!isVisible()) return;
    
    Theme* t = currentTheme();
    
    // Draw background
    Color bgColor = backgroundColor_;
    if (bgColor.a == 0) {
        bgColor = t ? t->colors.surface : Color::white();
    }
    
    if (cornerRadius_ > 0) {
        ctx->fillRoundedRect(bounds_, bgColor, cornerRadius_);
    } else {
        ctx->fillRect(bounds_, bgColor);
    }
    
    // Draw border
    Color borderColor = t ? t->colors.border : Color(0xCCCCCC);
    if (isFocused()) {
        borderColor = t ? t->colors.primary : Color(0x2196F3);
    }
    if (cornerRadius_ > 0) {
        ctx->strokeRoundedRect(bounds_, borderColor, cornerRadius_, 1.0f);
    } else {
        ctx->strokeRect(bounds_, borderColor, 1.0f);
    }
    
    // Clip to text area
    Rect textArea = bounds_.inset(padding_.left, padding_.top);
    ctx->pushClip(textArea);
    
    // Determine what to draw
    std::string displayText;
    Color displayColor;
    
    if (value_.empty() && !isFocused()) {
        displayText = placeholder_;
        displayColor = placeholderColor_;
    } else if (isPassword_) {
        displayText = std::string(value_.length(), '*');
        displayColor = textColor_;
    } else {
        displayText = value_;
        displayColor = textColor_;
    }
    
    // Draw selection highlight
    if (selectionStart_ != selectionEnd_ && isFocused()) {
        int start = std::min(selectionStart_, selectionEnd_);
        int end = std::max(selectionStart_, selectionEnd_);
        
        std::string beforeSel = value_.substr(0, start);
        std::string selected = value_.substr(start, end - start);
        
        float startX = ctx->measureText(beforeSel, 14.0f).width;
        float selWidth = ctx->measureText(selected, 14.0f).width;
        
        Color selColor = t ? t->colors.primary.withAlpha(60) : Color(33, 150, 243, 60);
        ctx->fillRect(Rect(textArea.x + startX, textArea.y, selWidth, textArea.height), selColor);
    }
    
    // Draw text
    float textY = textArea.y + textArea.height / 2 + 4;
    ctx->drawText(displayText, textArea.x, textY, displayColor, 14.0f);
    
    // Draw cursor
    if (isFocused() && cursorVisible_) {
        std::string beforeCursor = value_.substr(0, cursorPos_);
        float cursorX = textArea.x + ctx->measureText(beforeCursor, 14.0f).width;
        ctx->drawLine(cursorX, textArea.y + 2, cursorX, textArea.y + textArea.height - 2,
                      textColor_, 1.0f);
    }
    
    ctx->popClip();
}

Size TextField::measure(const Size& available) {
    (void)available;
    return Size(200, 36);  // Default size
}

EventResult TextField::onMouseDown(float x, float y, MouseButton button) {
    if (button != MouseButton::Left) return EventResult::Ignored;
    
    // Calculate cursor position from click
    float textX = bounds_.x + padding_.left;
    float clickOffset = x - textX;
    
    // Simple: approximate character position
    float charWidth = 8.0f;  // Approximate
    int pos = static_cast<int>(clickOffset / charWidth);
    pos = std::clamp(pos, 0, static_cast<int>(value_.length()));
    
    cursorPos_ = pos;
    selectionStart_ = pos;
    selectionEnd_ = pos;
    
    return EventResult::NeedsRedraw;
}

EventResult TextField::onKeyDown(KeyCode key, Modifiers mods) {
    if (isReadOnly_) return EventResult::Ignored;
    
    switch (key) {
        case KeyCode::Backspace:
            if (selectionStart_ != selectionEnd_) {
                deleteSelection();
            } else if (cursorPos_ > 0) {
                value_.erase(cursorPos_ - 1, 1);
                cursorPos_--;
                if (changeHandler_) changeHandler_(value_);
            }
            return EventResult::NeedsRedraw;
            
        case KeyCode::Delete:
            if (selectionStart_ != selectionEnd_) {
                deleteSelection();
            } else if (cursorPos_ < static_cast<int>(value_.length())) {
                value_.erase(cursorPos_, 1);
                if (changeHandler_) changeHandler_(value_);
            }
            return EventResult::NeedsRedraw;
            
        case KeyCode::Left:
            moveCursor(-1, mods.shift);
            return EventResult::NeedsRedraw;
            
        case KeyCode::Right:
            moveCursor(1, mods.shift);
            return EventResult::NeedsRedraw;
            
        case KeyCode::Home:
            cursorPos_ = 0;
            if (!mods.shift) selectionStart_ = selectionEnd_ = cursorPos_;
            return EventResult::NeedsRedraw;
            
        case KeyCode::End:
            cursorPos_ = static_cast<int>(value_.length());
            if (!mods.shift) selectionStart_ = selectionEnd_ = cursorPos_;
            return EventResult::NeedsRedraw;
            
        case KeyCode::Enter:
            if (submitHandler_) submitHandler_(value_);
            return EventResult::Handled;
            
        case KeyCode::A:
            if (mods.ctrl) {
                selectAll();
                return EventResult::NeedsRedraw;
            }
            break;
            
        default:
            break;
    }
    
    return EventResult::Ignored;
}

EventResult TextField::onTextInput(const std::string& text) {
    if (isReadOnly_) return EventResult::Ignored;
    insertText(text);
    return EventResult::NeedsRedraw;
}

EventResult TextField::onFocus() {
    state_.focused = true;
    cursorVisible_ = true;
    return EventResult::NeedsRedraw;
}

EventResult TextField::onBlur() {
    state_.focused = false;
    selectionStart_ = selectionEnd_ = 0;
    return EventResult::NeedsRedraw;
}

void TextField::insertText(const std::string& text) {
    if (selectionStart_ != selectionEnd_) {
        deleteSelection();
    }
    
    value_.insert(cursorPos_, text);
    cursorPos_ += static_cast<int>(text.length());
    
    if (maxLength_ > 0 && static_cast<int>(value_.length()) > maxLength_) {
        value_ = value_.substr(0, maxLength_);
        cursorPos_ = std::min(cursorPos_, maxLength_);
    }
    
    if (changeHandler_) changeHandler_(value_);
}

void TextField::deleteSelection() {
    int start = std::min(selectionStart_, selectionEnd_);
    int end = std::max(selectionStart_, selectionEnd_);
    value_.erase(start, end - start);
    cursorPos_ = start;
    selectionStart_ = selectionEnd_ = cursorPos_;
    if (changeHandler_) changeHandler_(value_);
}

void TextField::moveCursor(int delta, bool extend) {
    cursorPos_ = std::clamp(cursorPos_ + delta, 0, static_cast<int>(value_.length()));
    if (extend) {
        selectionEnd_ = cursorPos_;
    } else {
        selectionStart_ = selectionEnd_ = cursorPos_;
    }
}

} // namespace NXRender
